### ISSUE TITLE
VPAT Reader: announcements

### ISSUE THREAD

**Issue by abaevbog** [zotero/zotero#4223]
- [x] vpat 72: Show annotations/thumbnails/document outline buttons should announce that the sidebar view has changed after click.
- [x] vpat 73: Nothing is announced during keyboard navigation of the thumbnails in the sidebar. Similar to Preview, perhaps we should announce the page number of the thumbnail?
- [x] vpat 75: When the annotation are navigated with arrows in the scrollable content pane, there is no audible feedback. We should announce something to indicate what annotation has been focused and for the user to to not loose orientation within the document.
- [x] vpat 78: Similar to vpat 75, while navigating search results after typing something in "Find in document" (via enter or by clicking arrow buttons), we should include some kind of indication as to what is the current search result and where it is in the document.

**Comment by abaevbog** [comment]
@mrtcode so a couple of issues here that I did not touch in https://github.com/zotero/reader/pull/127 because they are a bit trickier are vpat 75 and 78. I thought I should check in with you on the best way to approach this (and if it's something I should approach to begin with).

vpat 75 and 78 are about how the selection within the reader can move between annotations or search results but it does not announce anything to screen reader.
Here is a [brief video](https://www.dropbox.com/scl/fi/840it2anqb0jiz0alzf3s/reader_announcements.mov?rlkey=sddsdr8pa74r6y7ehv3dljqbj&st=zc1m8wpf&dl=0) to display exactly what I mean. This gets interesting because, as far as I understand, the annotations or search results are never actually focused, just highlighted. So I was thinking maybe we could add a hidden [aria-live](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) region where we would insert the string to announce when the selection changes? That way things can get announced regardless of focus and such.

Do you think this makes sense for me to try? If so, where could I look in the reader to find where the highlighting of search results and keyboard selection of annotations in the actual reader happens?

**Comment by mrtcode** [comment]
We can try to do this. But where that aria-live region should be placed?

**Comment by abaevbog** [comment]
I don't believe there are any strict requirements as to where in the DOM they need to be. Overall, any node can be marked as `aria-live` and screen readers should be able to see and act on it.

Firefox has a similar node used exclusively for announcements and it is places just at the bottom of [browser.xhtml](https://searchfox.org/mozilla-central/source/browser/base/content/browser.xhtml#175). I think something like that would make sense?


If we add `a11tyAnnouncement ` node to the bottom of `index.reader.html` (script to read out on a loop "Count: 1", "Count: 2"... just for testing), the messages get properly announced on all screen readers that I tried.

So it would be just a matter of updating that node with proper content.

```
<!DOCTYPE html>
<html lang="en" dir="ltr" style="--sidebar-width: 240px; --split-view-size: 50%; --bottom-placeholder-height: 50px; --toolbar-placeholder-width: 0px">
<head>
	<meta charset="UTF-8">
	<title>Zotero Reader</title>
	<% if (build === "zotero") { %>
	<script>delete window.performance</script>
	<script src="resource://zotero/react.js"></script>
	<script src="resource://zotero/react-dom.js"></script>
	<script src="resource://zotero/react-intl.js"></script>
	<script src="resource://zotero/prop-types.js"></script>
	<% } %>
	<script>
		let count = 0;
		let announce = function() {
			setTimeout(() => {
				count += 1;
				document.querySelector("#a11tyAnnouncement").textContent = `Count: ${count}`;
				announce();
			}, 500);
		}
		announce();
	</script>
</head>
<body class="sidebar-open">
	<div id="reader-ui"></div>
	<div id="split-view">
		<div id="primary-view" class="primary"></div>
		<div id="secondary-view" class="secondary"></div>
	</div>
	<div id="printContainer"></div>
	<div id="a11tyAnnouncement" aria-live="polite"></div>
</body>
</html>
```

**Comment by mrtcode** [comment]
Thanks for the example. I’m wondering if we could place it inside the PDF view iframe, since it’s not needed for EPUB and snapshot views. This also means there would be two areas like this when using split view.

**Comment by abaevbog** [comment]
@mrtcode Apologies for late reply! I was moving last weekend and had to drive from one coast of the United States to another...

> since it’s not needed for EPUB and snapshot views

While annotations do not get selected with snapshots and EPUB (as far as I can tell), VPAT review indicates that we should also have some kind of announcement while going through search results. So we would want to support this (fairly similar) use case as well...

I was playing around with this and put together a draft suggestion on how it might work: https://github.com/zotero/reader/pull/130. It does need some tweaking to make localization messages a bit nicer and such but the main idea is that as you switch between annotations and search results, announcements do go through for PDFs and snapshots for me. This is what it looks like: [demo](https://www.dropbox.com/scl/fi/d0gdzcq0wj197sqdg7mw1/reader-aria-live.mov?rlkey=8eo3mbeh7x0u41xbdh5piyjor&st=8zkbn7s5&dl=0).

What do you think about this general approach?

**Comment by mrtcode** [comment]
That sounds right to me. You can merge your changes, and I’ll make a few tweaks.

**Comment by abaevbog** [comment]
> That sounds right to me. You can merge your changes, and I’ll make a few tweaks.

Cool! I just did with a minor cleanup and a bit more helpful messages.

**Comment by mrtcode** [comment]
Great! You can merge those.

**Comment by abaevbog** [comment]
Awesome! Hopefully (fingers crossed), with this we will have all the most pertinent reader issues covered!
